twenty-first/src/util_types/simple_hasher.rs

```diff
@@ -337,55 +332,66 @@ impl RescuePrimeXlix<RP_DEFAULT_WIDTH> {
     }
 }
 
-#[cfg(test)]
-pub mod test_simple_hasher {
-    use super::*;
-    use crate::shared_math::traits::GetRandomElements;
-    use rand::Rng;
+pub trait SamplableFrom<Digest> {
+    fn sample(digest: &Digest) -> Self;
+}
 
-    #[test]
-    fn u128_to_digest_test() {
-        let one = 1u128;
-        let bfields_one: Vec<BFieldElement> = one.to_digest();
-        assert_eq!(6, bfields_one.len());
-        assert_eq!(BFieldElement::ring_zero(), bfields_one[0]);
-        assert_eq!(BFieldElement::ring_zero(), bfields_one[1]);
-        assert_eq!(BFieldElement::ring_zero(), bfields_one[2]);
-        assert_eq!(BFieldElement::ring_zero(), bfields_one[3]);
-        assert_eq!(BFieldElement::ring_zero(), bfields_one[4]);
-        assert_eq!(BFieldElement::ring_one(), bfields_one[5]);
-
-        let beyond_bfield0 = u64::MAX as u128;
-        let bfields: Vec<BFieldElement> = beyond_bfield0.to_digest();
-        assert_eq!(6, bfields.len());
-        assert_eq!(BFieldElement::ring_zero(), bfields[0]);
-        assert_eq!(BFieldElement::ring_zero(), bfields[1]);
-        assert_eq!(BFieldElement::ring_zero(), bfields[2]);
-        assert_eq!(BFieldElement::ring_zero(), bfields[3]);
-        assert_eq!(BFieldElement::ring_one(), bfields[4]);
-        assert_eq!(BFieldElement::new(4294967295u64), bfields[5]);
-
-        let beyond_bfield1 = BFieldElement::MAX as u128 + 1;
-        let bfields: Vec<BFieldElement> = beyond_bfield1.to_digest();
-        assert_eq!(6, bfields.len());
-        assert_eq!(BFieldElement::ring_zero(), bfields[0]);
-        assert_eq!(BFieldElement::ring_zero(), bfields[1]);
-        assert_eq!(BFieldElement::ring_zero(), bfields[2]);
-        assert_eq!(BFieldElement::ring_zero(), bfields[3]);
-        assert_eq!(BFieldElement::ring_one(), bfields[4]);
-        assert_eq!(BFieldElement::ring_one(), bfields[5]);
-
-        let big_value = u128::MAX;
-        let bfields: Vec<BFieldElement> = big_value.to_digest();
-        for i in 1..128 {
-            let other_digest: Vec<BFieldElement> = (big_value >> i).to_digest();
-            assert_ne!(
-                bfields, other_digest,
-                "No digest collission allowed. i = {}",
-                i
-            );
+impl SamplableFrom<Vec<u8>> for BFieldElement {
+    fn sample(digest: &Vec<u8>) -> Self {
+        assert!(
+            digest.len() >= 8,
+            "Cannot sample pseudo-uniform BFieldElements from less than 8 bytes."
+        );
+        let mut integer = 0u64;
+        digest[0..8].iter().for_each(|d| {
+            integer *= 256u64;
+            integer += *d as u64;
+        });
+        BFieldElement::new(integer)
+    }
+}
```

Comment:
Prone to collisions since `BFieldElement::new(x) == BFieldElement::new(x + Q)`.

Reply:
The issue here is not collisions but non-uniformity. However, it _is_ an issue impacting security with a corresponding attack.

Reply:
Resolved in [0da29bd](https://github.com/Neptune-Crypto/twenty-first/pull/25/commits/0da29bd2934d6b85ddf652ea936f34d59bea560c).